### PR TITLE
KFSPTS-20851 adding webservice to KFS to support einvoice pagination

### PR DESCRIPTION
**Pull request by jhulslander**

### CODE REVIEW

**Review by ajd299 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResource.java [hunk 1]
```diff
@@ -103,6 +93,35 @@ public Response getVendors(@Context HttpHeaders headers) {
...
+            return Response.ok(buildVendorNumberList(vendorNumbers)).build();
+        } catch (Exception ex) {
+            LOG.error("getVendorNumbersSearchCriteria", ex);
+            return ex instanceof BadRequestException ? respondGetVendorBadRequest() : respondInternalServerError(ex);
+        }
+    }
+    
+    private String buildVendorNumberList(List<String> vendorNumbers) {
+        StringBuilder sb = new StringBuilder();
+        //String singleQuote = "'";
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResource.java:115]: yes

File: src/main/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResource.java [hunk 2]
```diff
@@ -103,6 +93,35 @@ public Response getVendors(@Context HttpHeaders headers) {
...
+        for (String vendorNumber : vendorNumbers) {
+            if (sb.length() > 0) {
+                sb.append(KFSConstants.COMMA);
+            }
+            sb.append(vendorNumber);
+        }
+        String vendorList =  sb.toString();
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResource.java:122]: yes, I will switch

File: src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java [hunk 3]
```diff
@@ -2,31 +2,88 @@
...
+    protected String buildSQl() {
+        StringBuilder sb = new StringBuilder("SELECT DETAIL.VNDR_HDR_GNRTD_ID, DETAIL.VNDR_DTL_ASND_ID ");
+        sb.append("FROM KFS.PUR_VNDR_DTL_T DETAIL, KFS.PUR_VNDR_DTL_TX EXTENSION ");
+        sb.append("WHERE DETAIL.VNDR_HDR_GNRTD_ID = EXTENSION.VNDR_HDR_GNRTD_ID ");
+        sb.append("AND DETAIL.VNDR_DTL_ASND_ID = EXTENSION.VNDR_DTL_ASND_ID ");
+        sb.append("AND EXTENSION.EINV_IND = '").append(EinvoiceIndicatorValuesFinder.EinvoiceIndicator.WEB.code).append("' ");
+        sb.append("AND (UPPER(DETAIL.VNDR_NM) LIKE ? ");
+        sb.append("OR UPPER(DETAIL.VNDR_DUNS_NBR) LIKE ?)");
+        String sql = sb.toString();
+        LOG.debug("buildSQl, generated SQL: " + sql);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java:84]: Yes, I think it is helpful, it is a debug log statement, so it won't normally print

**Review by ajd299 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/cornell/kfs/module/purap/dataaccess/CuEinvoiceDao.java [hunk 4]
```diff
@@ -6,4 +6,6 @@
 
 public interface CuEinvoiceDao {
     List<VendorDetail> getVendors(List<String> vendorNumbers);
+    
+    List<String> getFilteredVendorNumbers(String searchCritiera);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/dataaccess/CuEinvoiceDao.java:10]: ok

File: src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java [hunk 5]
```diff
@@ -2,31 +2,90 @@
...
+    @Override
     public List<VendorDetail> getVendors(List<String> vendorNumbers) {
         HashMap<Integer, List<Integer>> vendorDetailToHeaderMap = buildVendorDetailMap(vendorNumbers);
         Criteria criteria = buildQueryCriteria(vendorDetailToHeaderMap);
         Query query = QueryFactory.newQuery(VendorDetail.class, criteria);
         return (List<VendorDetail>) getPersistenceBrokerTemplate().getCollectionByQuery(query);
     }
+    
+    @Override
+    public List<String> getFilteredVendorNumbers(String searchCritiera) {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java:40]: ok

File: src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java [hunk 6]
```diff
@@ -2,31 +2,90 @@
...
+            }
+        } catch (SQLException e) {
+            LOG.error("getFilteredVendorNumbers, had an error processing the query results.", e);
+            throw new RuntimeException(e);
+        }
+        
+        return vendorNumbers;
+    }
+
+    protected ResultSet queryVendors(String searchCritiera) {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java:59]: ok

File: src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java [hunk 7]
```diff
@@ -2,31 +2,90 @@
...
+        ResultSet queryResults;
+        try {
+            Connection connection = getPersistenceBroker(true).serviceConnectionManager().getConnection();
+            PreparedStatement query = connection.prepareCall(buildSQl());
+            String formattedSearchCriteria = KFSConstants.PERCENTAGE_SIGN +  StringUtils.upperCase(searchCritiera).trim() + KFSConstants.PERCENTAGE_SIGN;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java:64]: ok

File: src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java [hunk 8]
```diff
@@ -2,31 +2,90 @@
...
+            query.setString(2, formattedSearchCriteria);
+            queryResults = query.executeQuery();
+        } catch (DataAccessResourceFailureException | LookupException | IllegalStateException | SQLException e) {
+            LOG.error("queryVendors, there was an error querying the database: ", e);
+            throw new RuntimeException(e);
+        }
+        return queryResults;
+    }
+    
+    protected String buildSQl() {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/dataaccess/impl/CuEinvoiceDaoOjb.java:75]: how about buildParameterizedVendorNumberQueryByVendorNameAndDunsNumber

File: src/test/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResourceTest.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class CuEinvoiceApiResourceTest {
+    
+    CuEinvoiceApiResource eInvoiceResource;
+
+    @BeforeEach
+    void setUp() throws Exception {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResourceTest.java:18]: nope, I shall remove

File: src/main/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResource.java [hunk 10]
```diff
@@ -103,6 +93,30 @@ public Response getVendors(@Context HttpHeaders headers) {
             return ex instanceof BadRequestException ? respondGetVendorsBadRequest() : respondInternalServerError(ex);
         }
     }
+    
+    @GET
+    @Path("vendornumbers/search/{searchcriteria}")
+    public Response getVendorNumbersSearchCriteria(@PathParam("searchcriteria") String searchcriteria, @Context HttpHeaders headers) {
+        LOG.debug("getVendorNumbersSearchCriteria, searchcriteria: " + searchcriteria);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/rest/resource/CuEinvoiceApiResource.java:100]: ok

**Comment by ajd299** [comment]
Code looks good, ready to merge once API is finalized

**Review by ajd299 — APPROVED** [review]

**Pull request merged**